Add Suspense page to docs (#11098)
apollographql/apollo-client: docs/source/data/suspense.mdx

Author: Updated description

---
title: Suspense
description: Using React 18 concurrent features with Apollo Client
minVersion: 3.8.0
---

"Suspense" is both a specific React API ([`<Suspense />`](https://react.dev/reference/react/Suspense), a component which "lets you display a fallback until its children have finished loading"), and often used more generally to refer to a new way to build React apps using the [**concurrent rendering engine**](https://react.dev/blog/2022/03/29/react-v18#what-is-concurrent-react) introduced in React 18.

> In this guide, we'll look at Apollo Client's new data fetching hooks introduced in v3.8—specifically `useSuspenseQuery`, `useBackgroundQuery` and `useReadQuery`—which leverage React's powerful concurrent features and walk through some examples of how they can be used.

Author: 👍 made a slightly different edit that also captures this

> To follow along with the examples below, open up our [Suspense demo](https://codesandbox.io/s/github/apollographql/docs-examples/tree/main/apollo-client/v3/suspense-hooks?file=/src/App.tsx) on CodeSandbox.

## Fetching with Suspense

> `useSuspenseQuery` initiates a network request and causes the component calling it to suspend while the request is in flight. It can be thought of as a drop-in replacement for `useQuery` that allows you to take advantage of React's concurrent features while fetching during render.

Reviewer: "suspenseful replacement" or "suspense-ready replacement" instead of "drop-in-replacement"?

> `useSuspenseQuery` initiates a network request and causes the component calling it to suspend while the request is in flight. It can be thought of as a drop-in replacement for `useQuery` that allows you to take advantage of React's concurrent features while fetching during render.

Author: I really like that @phryneas - updated

> `useSuspenseQuery` initiates a network request and causes the component calling it to suspend while the request is in flight. It can be thought of as a drop-in replacement for `useQuery` that allows you to take advantage of React's concurrent features while fetching during render.

Reviewer 2: Ya I like that a lot!

More details on `useSuspenseQuery`'s API can be found in [its API docs](../api/react/hooks/#usesuspensequery-api).

> > **Note:** a component using `useSuspenseQuery` to fetch data will **suspend on first render** unless it can access all of the data for the fields it is querying via the cache (or is configured with `returnPartialData: true` and partial data is present in the cache), so it **may not render until its initial network request finishes**. Keep this in mind and avoid creating "request waterfalls" whereby a series of components fetch data, suspend, receive data and render children which then initiate new network requests causing a "waterfall" of suspense boundaries to display. In some scenarios this can be avoided by using `useBackgroundQuery` and `useReadQuery` (see below).

Reviewer 2: Love the updates you made here. Thanks!

```tsx
import { Suspense } from 'react';
import { gql, TypedDocumentNode, useSuspenseQuery } from '@apollo/client';

> interface Data {

Author: Updated

  dog: {
    id: string;
    name: string;
  };
}

interface Variables {
  name: string;
}

const GET_DOG_QUERY: TypedDocumentNode<Data, Variables> = gql`
  query GetDog($name: String) {
    dog(name: $name) {
      id
      name
    }
  }
`;

function App() {
  return (
    <Suspense fallback={<div>Loading...</div>}>
      <Dog name="Mozzarella" />
    </Suspense>
  );
}

> function Dog({ name }: { name: string }) {

Author: 👍 Updated

  const { data } = useSuspenseQuery(GET_DOG_QUERY, {
    variables: { name },
  });

  return <>Name: {data.dog.name}</>;

> }

Author: Yeah fair point, this was based off of an existing example and I wanted to show the dog's name up front while there also aren't that many fields on `Dog`, I'll search by `id` instead.

> }

Reviewer: Maybe an easier change: query for `age` or `nickname`.

```

> In this example, our `App` component renders a `Dog` component which fetches the record for a single dog via `useSuspenseQuery`. When React attempts to render `Dog` for the first time, Apollo's `InMemoryCache` is empty, so `useSuspenseQuery` initiates a network request. `Dog` suspends while the network request is pending, triggering the nearest `Suspense` boundary _above_ the suspending component in `App` which renders our "Loading..." fallback. Once the network request is complete, `Dog` renders with the newly cached `name` for Mozzarella the Corgi.

Author: Hmm, I think I'll leave this one as is: with the default fetch policy, this data is cached so I think it's useful to indicate that.

> You may have noticed that `useSuspenseQuery` does not return a `loading` boolean: this is because the component calling `useSuspenseQuery` will always suspend when fetching data. A corollary is that when it _does_ render, `data` is always defined! With React 18's concurrent features, Suspense fallbacks that exist **outside of suspending components** take the place of loading states that components were responsible for rendering themselves in the previous React paradigm. When using concurrent React features, `Suspense` boundaries should be used throughout your application to render loading UI while components are suspended. The placement of suspense boundaries is a key part of designing a good user experience under React's concurrent rendering model.

Reviewer 2: Looking over this again, it sticks out too much so ignore this suggestion.

> You may have noticed that `useSuspenseQuery` does not return a `loading` boolean: this is because the component calling `useSuspenseQuery` will always suspend when fetching data. A corollary is that when it _does_ render, `data` is always defined! With React 18's concurrent features, Suspense fallbacks that exist **outside of suspending components** take the place of loading states that components were responsible for rendering themselves in the previous React paradigm. When using concurrent React features, `Suspense` boundaries should be used throughout your application to render loading UI while components are suspended. The placement of suspense boundaries is a key part of designing a good user experience under React's concurrent rendering model.

Reviewer 2: Meh, reading this again a day later doesn't seem like it adds much value. Ignore this one.

> You may have noticed that `useSuspenseQuery` does not return a `loading` boolean: this is because the component calling `useSuspenseQuery` will always suspend when fetching data. A corollary is that when it _does_ render, `data` is always defined! With React 18's concurrent features, Suspense fallbacks that exist **outside of suspending components** take the place of loading states that components were responsible for rendering themselves in the previous React paradigm. When using concurrent React features, `Suspense` boundaries should be used throughout your application to render loading UI while components are suspended. The placement of suspense boundaries is a key part of designing a good user experience under React's concurrent rendering model.

Reviewer 2: I like the edits that were made to this section, so I don't feel like this suggestion applies anymore. Ignore.

> You may have noticed that `useSuspenseQuery` does not return a `loading` boolean: this is because the component calling `useSuspenseQuery` will always suspend when fetching data. A corollary is that when it _does_ render, `data` is always defined! With React 18's concurrent features, Suspense fallbacks that exist **outside of suspending components** take the place of loading states that components were responsible for rendering themselves in the previous React paradigm. When using concurrent React features, `Suspense` boundaries should be used throughout your application to render loading UI while components are suspended. The placement of suspense boundaries is a key part of designing a good user experience under React's concurrent rendering model.

Reviewer 2: Looks like this was addressed. Thanks!

> > **Note for TypeScript users**: since `GET_DOG_QUERY` is a `TypedDocumentNode` in which we have specified the result type via `Data` generic, the TypeScript type for `data` returned by `useSuspenseQuery` reflects that! This means that `data` is guaranteed to be defined when `Dog` renders, and that `data.dog` will have the shape `{ id: string; name: string; breed: string; }`.

Reviewer 2: Everything reads pretty nicely and after coming back to this suggestion with the codegen mention in place, I'm happy with where this ended up.

### Changing variables

> In the previous example, we fetched the record for a single dog by passing a hard-coded `name` variable, Mozzarella, to `useSuspenseQuery`. Now, let's say we want to fetch the record for a different dog using a dynamic value instead. We'll fetch the list of dogs with just their `name` and `id`, and once the user selects an individual dog we can fetch more details, including their `breed`.

Author: I'll admit my personal preference is definitely for the way I wrote it - `a single dog record` reads strange to me - but maybe @Meschreiber has suggestions here :)

> In the previous example, we fetched the record for a single dog by passing a hard-coded `name` variable, Mozzarella, to `useSuspenseQuery`. Now, let's say we want to fetch the record for a different dog using a dynamic value instead. We'll fetch the list of dogs with just their `name` and `id`, and once the user selects an individual dog we can fetch more details, including their `breed`.

Reviewer 2: Ya you're right. Reading my suggestion a day later, I could take or leave the part about "Let's update our example". I'll leave this for now.

Let's take a look at some code:

```tsx
export const GET_DOG_QUERY: TypedDocumentNode<DogData, Variables> = gql`
  query GetDog($name: String) {
    dog(name: $name) {
      id
      name
      breed
    }
  }
`;

export const GET_DOGS_QUERY: TypedDocumentNode<DogsData, Variables> = gql`
  query GetDogs {
    dogs {
      id
      name
    }
  }
`;

function App() {
  const { data } = useSuspenseQuery(GET_DOGS_QUERY);
  const [selectedDog, setSelectedDog] = useState(data.dogs[0].name);

>   return (

Author: Updated everywhere

    <>
      <select
        onChange={(e) => {
          setSelectedDog(e.target.value);
        }}
      >
        {data.dogs.map((dog) => (
          <option value={dog.name}>{dog.name}</option>
        ))}
      </select>
      <Suspense fallback={<div>Loading...</div>}>
        <Dog name={selectedDog} />
      </Suspense>
    </>
  );
}

function Dog({ name }: { name: string }) {
  const { data } = useSuspenseQuery(GET_DOG_QUERY, {
    variables: { name },
  });
  return (
    <>
      <div>Name: {data.dog.name}</div>
      <div>Breed: {data.dog.breed}</div>
    </>
  );
}
```

> Changing the dog via the `select` will cause the component to suspend each time we select a dog whose record **does not yet exist in the cache**. Once we've loaded a given dog's record into Apollo Client's `InMemoryCache`, however, selecting that dog again from our dropdown will not cause the component to re-suspend, since under our default `cache-first` fetch policy Apollo Client will not make a network request after a cache hit.

Author: Good point - believe we also do this in the fragment docs (use InMemoryCache as the shorthand reference for the cache since it's the most common case)

> Changing the dog via the `select` will cause the component to suspend each time we select a dog whose record **does not yet exist in the cache**. Once we've loaded a given dog's record into Apollo Client's `InMemoryCache`, however, selecting that dog again from our dropdown will not cause the component to re-suspend, since under our default `cache-first` fetch policy Apollo Client will not make a network request after a cache hit.

Reviewer 2: Coming back to this, I'll leave this for now and we can edit if @Meschreiber prefers something different 🙂

> Changing the dog via the `select` will cause the component to suspend each time we select a dog whose record **does not yet exist in the cache**. Once we've loaded a given dog's record into Apollo Client's `InMemoryCache`, however, selecting that dog again from our dropdown will not cause the component to re-suspend, since under our default `cache-first` fetch policy Apollo Client will not make a network request after a cache hit.

Reviewer 2: I like the edits made here. Thanks!

```tsx {10-12}
import { useState, Suspense, startTransition } from "react";

function App() {
  const { data } = useSuspenseQuery(GET_DOGS_QUERY);
  const [selectedDog, setSelectedDog] = useState(data.dogs[0].name);
  return (
    <>
      <select
        onChange={(e) => {
          startTransition(() => {
            setSelectedDog(e.target.value);
          });
        }}
      >
        {data.dogs.map((dog) => (
          <option value={dog.name}>{dog.name}</option>
        ))}
      </select>
      <Suspense fallback={<div>Loading...</div>}>
        <Dog name={selectedDog} />
      </Suspense>
    </>
  );
}
```

By wrapping our call to `setSelectedDog` in our `App` component in React's `startTransition` method, we no longer see the suspense fallback every time we select a new dog! Instead, the previous dog remains on the screen while the next dog's record is fetched in the background. Once the network request is complete, the new dog's record is immediately displayed.

> 

Reviewer 2: Moved this suggestion to a single comment.

The `Dog` component **does not suspend** while this request is being made. This is because we've opted to render partial data _instead_ of suspending by setting `returnPartialData: true`. Once the network request is complete, `Dog` re-renders with the `breed` now displayed.

> **Note for TypeScript users**: with `returnPartialData` to `true` in `UseSuspenseQueryOptions`, the type for the returned `data` still accurately reflects the shape of the result we expect, but all fields are now also possibly `undefined` since we don't know what data will exist in the cache on first render.

### Error handling

> In the concurrent React paradigm, errors thrown in React components are caught and displayed by the closest [error boundary](https://react.dev/reference/react/Component#static-getderivedstatefromerror) _on the client_. _On the server_, if you're use a streaming server rendering API or a framework built with one, [the behavior is slightly different](https://react.dev/reference/react/Suspense#providing-a-fallback-for-server-errors-and-client-only-content): the nearest `Suspense` boundary will be rendered by the server so the user initially sees a loading indicator, and the client will attempt to render the component again. If it succeeds, the user is never aware of the error thrown on the server. If it errors again on the client, the error boundary is displayed.

Author: Slightly tweaked this to make it clear that it's only when using streaming server rendering APIs, not just SSR in general.

> > **Note:** An error boundary is a **class component** that implements `static getDerivedStateFromError`. If you don't want to write an error boundary class component yourself, you can use a library like [`react-error-boundary`](https://github.com/bvaughn/react-error-boundary) instead. For more information, see the React docs for [catching rendering errors with an error boundary](https://react.dev/reference/react/Component#catching-rendering-errors-with-an-error-boundary).

Reviewer 2: After reading this again, I like the way it is.

  refetchHandler: () => void;
}) {
  const { data } = useSuspenseQuery(GET_DOG_QUERY, {
    variables: { name },
  });
  return (
    <>
      Name: {data.dog.name}
      <Suspense fallback={<div>Loading breeds...</div>}>
        <Breeds queryRef={queryRef} />
      </Suspense>
      <button onClick={refetchHandler}>Refetch!</button>
    </>
  );
}
```

In this example, every time the user clicks the "Refetch!" button rendered by the `Dog` component, our `refetchHandler` calls the `refetch` function returned by `useBackgroundQuery` which initiates a new network request for the `GET_BREEDS_QUERY` query. As a result, the `Breeds` component re-suspends while the new network request is pending.

## Avoiding request waterfalls

Since `useSuspenseQuery` will usually cause a component to suspend on first render while data is being fetched, a tree of components that all use `useSuspenseQuery` to query different data can cause a "waterfall" of suspense boundaries to display since each component will suspend before rendering its children (which will then suspend while fetching, and so on). In some cases, this can be avoided by using `useBackgroundQuery` and `useReadQuery` instead.

> `useBackgroundQuery` initiates a request for data in a parent component and returns a `QueryReference` which is used to read the data in a child component via `useReadQuery`. If the child component attempts to render before the data can be found in the cache, the child component will suspend until the data is available. On cache updates to watched data, the child component calling `useReadQuery` will re-render with new data **but the parent component will not re-render** (as it would, for example, if it were using `useQuery` to issue the request).

Author: 👍 made some updates here

> **Note:** in a client-rendered application, `useBackgroundQuery` can be used to avoid request waterfalls, but its impact can be even more noticeable in an application using streaming SSR ([when using Apollo Client's suspense hooks with Next.js's App Router](./#using-apollo-clients-suspense-hooks-with-nextjss-app-router), for example). This is because the server can begin streaming content to the client, bringing even greater performance benefits.

More details on `useBackgroundQuery` and `useReadQuery`'s API can be found in [their API docs](../api/react/hooks/#usebackgroundquery-and-usereadquery).

```tsx

> function App() {

Author: Updated

  const [queryRef] = useBackgroundQuery(GET_BREEDS_QUERY);
  return (
    <Suspense fallback={<div>Loading...</div>}>
      <Dog name="Mozzarella" queryRef={queryRef} />
    </Suspense>
  );
}

function Dog({
  name,
  queryRef,
}: {
  name: string;
  queryRef: QueryReference<BreedData>;
}) {
  const { data } = useSuspenseQuery(GET_DOG_QUERY, {
    variables: { name },
  });
  return (
    <>
      Name: {data.dog.name}
      <Suspense fallback={<div>Loading breeds...</div>}>
        <Breeds queryRef={queryRef} />
      </Suspense>
    </>
  );
}

function Breeds({ queryRef }: { queryRef: QueryReference<BreedData> }) {
  const { data } = useReadQuery(queryRef);
  return data.breeds.map(({ characteristics }) =>
    characteristics.map((characteristic) => (
      <div key={characteristic}>{characteristic}</div>
    ))
  );
}
```

> ## Using Apollo Client's Suspense Hooks with Next.js's App Router

Reviewer 2: To leave room for the future, should we consider making this section a bit more generic and having `Next.js` be a subsection of this one?

---

```suggestion
## React Server Components (RSC)

### Usage with Next.js 13 App Router
```

Thoughts? totally cool if we wait to make this change as well since we don't have another entry at the moment.

---
title: Suspense
description: Using React 18 Suspense features with Apollo Client
minVersion: 3.8.0
---

"Suspense" is both a specific React API ([`<Suspense />`](https://react.dev/reference/react/Suspense), a component which "lets you display a fallback until its children have finished loading"), and often used more generally to refer to a new way to build React apps using the [**concurrent rendering engine**](https://react.dev/blog/2022/03/29/react-v18#what-is-concurrent-react) introduced in React 18.

In this guide, we'll look at Apollo Client's data fetching hooks introduced in 3.8 which leverage React's powerful Suspense features and walk through some examples of how they can be used.

> To follow along with the examples below, open up our [Suspense demo](https://codesandbox.io/s/github/apollographql/docs-examples/tree/main/apollo-client/v3/suspense-hooks?file=/src/App.tsx) on CodeSandbox.

## Fetching with Suspense

> `useSuspenseQuery` initiates a network request and causes the component calling it to suspend while the request is being made. It can be thought of as a Suspense-ready replacement for `useQuery` that allows you to take advantage of React's Suspense features while fetching during render.

Reviewer 2: ```suggestion
`useSuspenseQuery` initiates a network request and causes the component calling it to suspend while the request is made. It can be thought of as a Suspense-ready replacement for `useQuery` that allows you to take advantage of React's Suspense features while fetching during render.
```

Let's take a look at an example:

```tsx {5,38,40,45-47}
import { Suspense } from 'react';
import {
  gql,
  TypedDocumentNode,
  useSuspenseQuery
} from '@apollo/client';

interface Data {
  dog: {
    id: string;
    name: string;
  };
}

interface Variables {
  id: string;
}

interface DogProps {
  id: string
}

const GET_DOG_QUERY: TypedDocumentNode<Data, Variables> = gql`
  query GetDog($id: String) {
    dog(id: $id) {
      # By default, an object's cache key is a combination of
      # its __typename and id fields, so we should always make
      # sure the id is in the response so our data can be

>       # properly cached.

Reviewer 2: Super weird. This shows up aligned in this doc, but looking at the preview, the comments are outdented too far.

<img width="640" alt="Screenshot 2023-08-04 at 2 56 48 PM" src="https://github.com/apollographql/apollo-client/assets/565661/8379de43-0f4d-497c-ac3d-75eb30757531">

Maybe there is a tab vs space or something in here?

Author: much better :)

![CleanShot 2023-08-04 at 17 12 47](https://github.com/apollographql/apollo-client/assets/5139846/c4ccc694-3816-48a1-9a7c-f35b3eebecc4)

      id
      name
    }
  }
`;

function App() {
  return (
    <Suspense fallback={<div>Loading...</div>}>
      <Dog id="3" />
    </Suspense>
  );
}

function Dog({ id }: DogProps) {
  const { data } = useSuspenseQuery(GET_DOG_QUERY, {
    variables: { id },
  });

  return <>Name: {data.dog.name}</>;
}
```

> **Note:** this example manually defines TypeScript interfaces for `Data` and `Variables` as well as the type for `GET_DOG_QUERY` using `TypedDocumentNode`. [GraphQL Code Generator](https://www.the-guild.dev/graphql/codegen) is a popular tool that will create these type definitions automatically for you. See the reference on [Using TypeScript](../development-testing/static-typing) for more information on integrating GraphQL Code Generator with Apollo Client.

> In this example, our `App` component renders a `Dog` component which fetches the record for a single dog via `useSuspenseQuery`. When React attempts to render `Dog` for the first time, the cache is empty so `useSuspenseQuery` initiates a network request. `Dog` suspends while the network request is pending, triggering the nearest Suspense boundary _above_ the suspending component in `App` which renders our "Loading..." fallback. Once the network request is complete, `Dog` renders with the newly cached `name` for Mozzarella the Corgi.

Reviewer 2: ```suggestion
In this example, our `App` component renders a `Dog` component which fetches the record for a single dog via `useSuspenseQuery`. When React attempts to render `Dog` for the first time, the cache is empty so `useSuspenseQuery` initiates a network request. `Dog` suspends while the network request is pending, triggering the nearest Suspense boundary _above_ the suspended component in `App` which renders our "Loading..." fallback. Once the network request is complete, `Dog` renders with the newly cached `name` for Mozzarella the Corgi.
```

You may have noticed that `useSuspenseQuery` does not return a `loading` boolean: this is because the component calling `useSuspenseQuery` will always suspend when fetching data. A corollary is that when it _does_ render, `data` is always defined! Suspense fallbacks that exist **outside of suspending components** take the place of loading states that components were responsible for rendering themselves in the previous React paradigm.

> > **Note for TypeScript users**: since `GET_DOG_QUERY` is a `TypedDocumentNode` in which we have specified the result type via `Data` generic, the TypeScript type for `data` returned by `useSuspenseQuery` reflects that! This means that `data` is guaranteed to be defined when `Dog` renders, and that `data.dog` will have the shape `{ id: string; name: string; breed: string; }`.

Reviewer 2: ```suggestion
> **Note for TypeScript users**: since `GET_DOG_QUERY` is a `TypedDocumentNode` in which we have specified the result type via `Data` generic argument, the TypeScript type for `data` returned by `useSuspenseQuery` reflects that! This means that `data` is guaranteed to be defined when `Dog` renders, and that `data.dog` will have the shape `{ id: string; name: string; breed: string; }`.
```

Saw this thread the other day and this made me think of this 🤣 https://twitter.com/mattpocockuk/status/1682030938790985736?s=20

  };
}

interface Variables {
  id: string;
}

interface DogProps {
  id: string
}

const GET_DOG_QUERY: TypedDocumentNode<Data, Variables> = gql`
  query GetDog($id: String) {
    dog(id: $id) {
      # By default, an object's cache key is a combination of
      # its __typename and id fields, so we should always make
      # sure the id is in the response so our data can be
      # properly cached.
      id
      name
    }
  }
`;

function App() {
  return (
    <Suspense fallback={<div>Loading...</div>}>
      <Dog id="3" />
    </Suspense>
  );
}

function Dog({ id }: DogProps) {
  const { data } = useSuspenseQuery(GET_DOG_QUERY, {
    variables: { id },
  });

  return <>Name: {data.dog.name}</>;
}
```

> **Note:** this example manually defines TypeScript interfaces for `Data` and `Variables` as well as the type for `GET_DOG_QUERY` using `TypedDocumentNode`. [GraphQL Code Generator](https://www.the-guild.dev/graphql/codegen) is a popular tool that will create these type definitions automatically for you. See the reference on [Using TypeScript](../development-testing/static-typing) for more information on integrating GraphQL Code Generator with Apollo Client.

> In this example, our `App` component renders a `Dog` component which fetches the record for a single dog via `useSuspenseQuery`. When React attempts to render `Dog` for the first time, the cache is empty so `useSuspenseQuery` initiates a network request. `Dog` suspends while the network request is pending, triggering the nearest Suspense boundary _above_ the suspended component in `App` which renders our "Loading..." fallback. Once the network request is complete, `Dog` renders with the newly cached `name` for Mozzarella the Corgi.

Reviewer 2: ```suggestion
In this example, our `App` component renders a `Dog` component which fetches the record for a single dog via `useSuspenseQuery`. When React attempts to render `Dog` for the first time, the cache is unable to fulfill the request for the `GetDog` query, so `useSuspenseQuery` initiates a network request. `Dog` suspends while the network request is pending, triggering the nearest `Suspense` boundary _above_ the suspended component in `App` which renders our "Loading..." fallback. Once the network request is complete, `Dog` renders with the newly cached `name` for Mozzarella the Corgi.
```

I would avoid saying that the cache is "empty" here and gear it more towards the fact the cache can't fulfill the data required for the query. In this way, it alludes to the fact that the cache can have other data in it, it just doesn't have enough to fulfill the request, so it suspends. Saying its "empty" might confuse someone who thinks that this is a precondition for suspense to work.

You may have noticed that `useSuspenseQuery` does not return a `loading` boolean: this is because the component calling `useSuspenseQuery` will always suspend when fetching data. A corollary is that when it _does_ render, `data` is always defined! Suspense fallbacks that exist **outside of suspended components** take the place of loading states that components were responsible for rendering themselves in the previous React paradigm.

> **Note for TypeScript users**: since `GET_DOG_QUERY` is a `TypedDocumentNode` in which we have specified the result type via `Data` generic type argument, the TypeScript type for `data` returned by `useSuspenseQuery` reflects that! This means that `data` is guaranteed to be defined when `Dog` renders, and that `data.dog` will have the shape `{ id: string; name: string; breed: string; }`.

### Changing variables

> In the previous example, we fetched the record for a single dog by passing a hard-coded `id` variable, Mozzarella, to `useSuspenseQuery`. Now, let's say we want to fetch the record for a different dog using a dynamic value instead. We'll fetch the list of dogs with just their `name` and `id`, and once the user selects an individual dog we can fetch more details, including their `breed`.

Reviewer 2: ```suggestion
In the previous example, we fetched the record for a single dog by passing a hard-coded `id` variable, Mozzarella, to `useSuspenseQuery`. Now, let's say we want to fetch the record for a different dog using a dynamic value. We'll fetch the list of dogs with just their `name` and `id`, and once the user selects an individual dog we can fetch more details, including their `breed`.
```

I think making this a touch more concise makes this read a little bit nicer.

> Let's take a look at some code:

Reviewer 2: ```suggestion
Let's update our example:
```

I like to think of us as guiding the user in their own editor through these examples. Rather than merely "looking" at code, I like telling the user to update the example. This is especially useful since you provided the codesandbox link above to follow along.

Author: Nice, I really like the impact of this subtle change 👍

      name
    }
  }
`;

function App() {
  const { data } = useSuspenseQuery(GET_DOGS_QUERY);
  const [selectedDog, setSelectedDog] = useState(
    data.dogs[0].id
  );

  return (
    <>
      <select
        onChange={(e) => setSelectedDog(e.target.value)}
      >
        {data.dogs.map(({ id, name }) => (
          <option key={id} value={id}>{dog.name}</option>
        ))}
      </select>
      <Suspense fallback={<div>Loading...</div>}>
        <Dog id={selectedDog} />
      </Suspense>
    </>
  );
}

function Dog({ id }: DogProps) {
  const { data } = useSuspenseQuery(GET_DOG_QUERY, {
    variables: { id },
  });

  return (
    <>
      <div>Name: {data.dog.name}</div>
      <div>Breed: {data.dog.breed}</div>
    </>
  );
}
```

Changing the dog via the `select` will cause the component to suspend each time we select a dog whose record **does not yet exist in the cache**. Once we've loaded a given dog's record in the cache, however, selecting that dog again from our dropdown will not cause the component to re-suspend, since under our default `cache-first` fetch policy Apollo Client will not make a network request after a cache hit.

### Updating state without suspending

> While we may sometimes want to suspend while we update some state that results in a pending network request, there are many cases when we would prefer to keep displaying the previous render while React prepares to update in the background.

Reviewer 2: ```suggestion
There are times that we want to avoid showing the loading UI in response to a pending network request and instead prefer to maintain the already rendered content. We can use a _transition_ to mark our update as non-urgent, which tells React to keep the existing UI in place until the new data has finished loading.
```

I went spelunking in the suspense docs and I actually didn't find mention of the word "background". This suggestion updates this to more align with the [suspense docs](https://react.dev/reference/react/Suspense) and the language used there. See if you like this edit.

Author: Yeah good call, just made some tiny adjustments and added a link to the blog post about transitions since that gives the best high level overview if anyone wants to read more.

Reviewer 2: Good call. Love it.

> In these cases, we can use a pair of React APIs that allow us to update state without suspending: [`startTransition`](https://react.dev/reference/react/startTransition)/[`useTransition`](https://react.dev/reference/react/useTransition) and [`useDeferredValue`](https://react.dev/reference/react/useDeferredValue). Please see the React documentation for a full set of usage examples and caveats.

Reviewer 2: ```suggestion
To mark a state update as a transition, we use the [`startTransition`](https://react.dev/reference/react/startTransition) function from React.

> For a deep-dive into these APIs, see the linked React documentation for a full set of usage examples and caveats.
```

To more align with the verbiage in the suspense docs, I think simplifying it to this makes sense and moving the mention of React documentation to a block quote. Let me know if you like this change.

> Let's modify our code so that the previously displayed dog remains on the screen while the next one is fetched in the background:

Reviewer 2: ```suggestion
Let's modify our example so that the previously displayed dog remains on the screen while the next one is fetched in the background:
```

Update to keep with the theme of "the example"

```tsx {1,13-15}
import { useState, Suspense, startTransition } from "react";

function App() {
  const { data } = useSuspenseQuery(GET_DOGS_QUERY);
  const [selectedDog, setSelectedDog] = useState(
    data.dogs[0].id
  );

  return (
    <>
      <select
        onChange={(e) => {
          startTransition(() => {
            setSelectedDog(e.target.value);
          });
        }}
      >
        {data.dogs.map(({ id, name }) => (
          <option key={id} value={id}>{name}</option>
        ))}
      </select>
      <Suspense fallback={<div>Loading...</div>}>
        <Dog id={selectedDog} />
      </Suspense>
    </>
  );
}
```

By wrapping our call to `setSelectedDog` in our `App` component in React's `startTransition` method, we no longer see the Suspense fallback every time we select a new dog! Instead, the previous dog remains on the screen while the next dog's record is fetched in the background. Once the network request is complete, the new dog's record is immediately displayed.

### Rendering partial data

> You may want to render your component immediately if partial data is available in the cache without suspending first. To do this, you can use the `returnPartialData` option **combined with one of two `fetchPolicy` settings**: `cache-first` (default) or `cache-and-network`.

Reviewer 2: ```suggestion
When the cache contains partial data, you may prefer to render that partial data immediately without suspending. To do this, you can use the `returnPartialData` option.

> **Note:** this option only works when combined with either the `cache-first` (default) or `cache-and-network` fetch policy. `cache-only` is not currently supported by `useSuspenseQuery`. For details on these fetch policies, see [Setting a fetch policy](/react/data/queries/#setting-a-fetch-policy).
```

You might consider moving the information about the fetch policy part of a block quote since its secondary information. I'd also combine information from the next block quote into here.

Here is a suggestion as well to re-arrange the sentence slightly which makes the "render immediate" a consequence of having partial data in the cache rather than the other way around. Doing so eliminates the "if" part of this statement.

> > **A note on the `fetchPolicy` option:** when fetching data with Apollo Client's Suspense hooks, the four supported fetch policies are `cache-first` (default), `network-only`, `no-cache` and `cache-and-network`. For details on these fetch policies, see [Setting a fetch policy](./queries/#setting-a-fetch-policy).

Reviewer 2: See https://github.com/apollographql/apollo-client/pull/11098/files#r1284880618. If you like that suggestion, you can remove this.

```suggestion
```

> Consider how we can adjust our example `App` and `Dog` components to render partial data on first render:

Reviewer 2: ```suggestion
Here's how we adjust our example to render partial data:
```

Since the user is following along, I'd prefer talking about making code modifications rather than just "considering" it.

```tsx {31}

> const GET_DOG_QUERY_PARTIAL: TypedDocumentNode = gql`

Reviewer 2: ```suggestion
const GET_PARTIAL_DOG_QUERY: TypedDocumentNode = gql`
```

I tend to prefer this style because "partial" feels more like an adjective to me than a noun. Switching this around makes the dog query the noun instead.

Then again, its a code sample, so not the end of the world if you prefer the way it is 😄

### `refetch`, `fetchMore` and `subscribeToMore`

Apollo's Suspense data fetching hooks return functions for re-initiating the same query (with existing or new variables) or to fetch more paginated data. `useSuspenseQuery` also provides a `subscribeToMore` function to execute a subscription.

When using `useBackgroundQuery`, both `refetch` and `fetchMore` functions can be destructured from the second item in the tuple returned by the hook.

Let's take a look at an example:

```tsx {1,3-5,12-15,18-20,49,57,62}
import { Suspense, useTransition } from "react";
import {
  useSuspenseQuery,
  useBackgroundQuery,
  useReadQuery,
  gql,
  TypedDocumentNode,
  QueryReference,
} from "@apollo/client";

function App() {

>   const [isPending, startTransition] = useTransition();

Reviewer 2: I love that you updated this to include this bit about `useTranstion`. Re-reading this, I think this would flow a bit better by moving this example right under the "[Updating state without suspending](https://deploy-preview-11098--apollo-client-docs.netlify.app/data/suspense/#updating-state-without-suspending)" section.

Here was my previous suggestion on moving this section up a touch:

---

### Showing pending UI during a transition

In the previous example, there is no visual indication that a fetch is happening when a new dog is selected. To provide nice visual feedback, let's update our example to use React's [`useTransition`](https://react.dev/reference/react/useTransition) hook which gives you an `isPending` boolean value to determine when a transition is happening. We'll dim the select dropdown while the transition is happening.

```tsx
import { useState, Suspense, useTransition } from "react";

function App() {
  const [isPending, startTransition] = useTransition();
  const { data } = useSuspenseQuery(GET_DOGS_QUERY);
  const [selectedDog, setSelectedDog] = useState(data.dogs[0].name);

  return (
    <>
      <select
        style={{ opacity: isPending ? 0.5 : 1 }}
        onChange={(e) => {
          startTransition(() => {
            setSelectedDog(e.target.value);
          });
        }}
      >
        {data.dogs.map((dog) => (
          <option value={dog.name}>{dog.name}</option>
        ))}
      </select>
      <Suspense fallback={<div>Loading...</div>}>
        <Dog name={selectedDog} />
      </Suspense>
    </>
  );
}
```

## Distinguishing between queries with `queryKey`

In rare cases, you might need to render multiple components that use the same `query` and `variables` combination. Due to the mechanics of Suspense, Apollo Client uses the combination of `query` and `variables` to uniquely identify the query when using Apollo's `Suspense` data fetching hooks.

This presents a problem, however, because it means multiple hooks may end up with the same identity. These hooks will suspend together, regardless of which component initiates or re-initiates a network request.

You can prevent this with `queryKey` option to ensure each hook has a unique identity. When `queryKey` is provided, Apollo Client will use as part of the hook's identity in combination with its `query` and `variables`. 

> For more information, see the [`useSuspenseQuery`](../api/react/hooks/#usesuspensequery) or [`useBackgroundQuery`](../api/react/hooks/#usebackgroundquery-and-usereadquery) API docs.

Author: The useBackgroundQuery link is broken here

Reviewer 2: I can get this fixed :)

Reviewer 2: [971a802](https://github.com/apollographql/apollo-client/pull/11098/commits/971a802a84ca2e79e94c84786a6a3076a19332d4)
